src/common/blkdev.cc

```diff
@@ -281,89 +253,263 @@ int get_block_device_size(int fd, int64_t *psize)
+}
+
+int BlkDev::get_block_device_size(int64_t *psize)
 {
   int ret = ::ioctl(fd, DIOCGMEDIASIZE, psize);
   if (ret < 0)
     ret = -errno;
   return ret;
 }
 
-bool block_device_support_discard(const char *devname)
+bool BlkDev::block_device_support_discard()
 {
-  return false;
+  struct diocgattr_arg arg;
+  int ret;
+
+  strlcpy(arg.name, "GEOM::candelete", sizeof(arg.name));
+  arg.len = sizeof(arg.value.i);
+  if (ioctl(fd, DIOCGATTR, &arg) == 0)
+    ret = (arg.value.i != 0);
+  else
+    ret = false;
+
+  return ret;
 }
 
-int block_device_discard(int fd, int64_t offset, int64_t len)
+int BlkDev::block_device_discard(int64_t offset, int64_t len)
 {
   return -EOPNOTSUPP;
 }
 
-bool block_device_is_rotational(const char *devname)
+bool BlkDev::block_device_is_nvme()
 {
-  return false;
+  // FreeBSD doesn't have a good way to tell if a device's underlying protocol
+  // is NVME, especially since multiple GEOM transforms may be involved.  So
+  // we'll just guess based on the device name.
+  struct fiodgname_arg arg;
+  const char *nda = "nda";        //CAM-based attachment
+  const char *nvd = "nvd";        //CAM-less attachment
+  char devname[PATH_MAX];
+
+  arg.buf = devname;
+  arg.len = sizeof(devname);
+  if (ioctl(fd, FIODGNAME, &arg) < 0)
+    return false; //When in doubt, it's probably not NVME
+
+  return (strncmp(nvd, devname, strlen(nvd)) == 0 ||
+          strncmp(nda, devname, strlen(nda)) == 0);
 }
 
-int get_device_by_uuid(uuid_d dev_uuid, const char* label, char* partition,
-	char* device)
+bool BlkDev::block_device_is_rotational()
 {
-  return -EOPNOTSUPP;
+#if __FreeBSD_version >= 1200049
+  struct diocgattr_arg arg;
+  int ioctl_ret;
+  bool ret;
+
+  strlcpy(arg.name, "GEOM::rotation_rate", sizeof(arg.name));
+  arg.len = sizeof(arg.value.u16);
+
+  ioctl_ret = ioctl(fd, DIOCGATTR, &arg);
+  if (ioctl_ret < 0 || arg.value.u16 == DISK_RR_UNKNOWN)
+    // DISK_RR_UNKNOWN usually indicates an old drive, which is usually spinny
+    ret = true;
+  else if (arg.value.u16 == DISK_RR_NON_ROTATING)
+    ret = false;
+  else if (arg.value.u16 >= DISK_RR_MIN && arg.value.u16 <= DISK_RR_MAX)
+    ret = true;
+  else
+    ret = true;     // Invalid value.  Probably spinny?
+
+  return ret;
+#else
+  return true;      // When in doubt, it's probably spinny
+#endif
 }
-int get_device_by_fd(int fd, char *partition, char *device, size_t max)
+
+int BlkDev::block_device_model(char *model, size_t max)
 {
-  return -EOPNOTSUPP;
+  struct diocgattr_arg arg;
+  char *p;
+
+  strlcpy(arg.name, "GEOM::descr", sizeof(arg.name));
+  arg.len = sizeof(arg.value.str);
+  if (ioctl(fd, DIOCGATTR, &arg) < 0) {
+    return -errno;
+  }
+
+  // The GEOM description is of the form "vendor product" for SCSI disks
+  // and "ATA device_model" for ATA disks.  Some vendors choose to put the
+  // vendor name in device_model, and some don't.  Strip the first bit.
+  p = arg.value.str;
+  (void) strsep(&p, " ");
+  snprintf(model, max, "%s", p);
```

Comment:
On my vm the ioctl succeeds but `arg.value.str` is empty. As a result `strsep` sets `p` to null:
```
Partition:      /dev/vtbd0p2
Model:          (null)
```
Probably it is better to return "Unknown" in this case? Or at least it is safer to check `p` is not null.